## 1. EDA — первинне дослідження даних Olist

Мета цього ноутбука — дослідити 9 CSV-файлів датасету **до будь-якого очищення**: зрозуміти структуру кожної таблиці, виявити проблеми з якістю даних і сформулювати рішення для ETL-етапу.

Перевіряємо чотири категорії:
- **Типи даних** — чи правильно pandas розпізнав дати, числа, рядки
- **Пропуски (NULL)** — де і скільки, чи це норма чи аномалія
- **Дублікати та унікальність** — перевірка первинних ключів, повні дублікати рядків
- **Цілісність зв'язків (FK)** — чи всі значення дочірніх таблиць мають відповідники в батьківських

Результат — зведена таблиця у секції 1.6: кожна знайдена проблема, її критичність і рішення, що буде реалізоване в `02_etl.ipynb`.

---

### 1.1 Завантаження даних

Читаємо всі 9 CSV-файлів у словник DataFrames. На цьому кроці — тільки завантаження без будь-яких трансформацій. Виводимо розмір кожного файлу для звірки з описом датасету.

In [1]:
import pandas as pd
from pathlib import Path

DATA_DIR = Path.cwd().parent / "data" / "raw"

files = {
    'customers': 'olist_customers_dataset.csv',
    'orders': 'olist_orders_dataset.csv',
    'order_items': 'olist_order_items_dataset.csv',
    'order_payments': 'olist_order_payments_dataset.csv',
    'order_reviews': 'olist_order_reviews_dataset.csv',
    'products': 'olist_products_dataset.csv',
    'sellers': 'olist_sellers_dataset.csv',
    'geolocation': 'olist_geolocation_dataset.csv',
    'category_translation': 'product_category_name_translation.csv',
}

dfs = {}
for key, path in files.items():
    dfs[key] = pd.read_csv(DATA_DIR / path)

# Швидка перевірка — скільки рядків у кожному файлі
for name, df in dfs.items():
    print(f"{name:25s} → {df.shape[0]:>7,} rows, {df.shape[1]:>2} cols")

customers                 →  99,441 rows,  5 cols
orders                    →  99,441 rows,  8 cols
order_items               → 112,650 rows,  7 cols
order_payments            → 103,886 rows,  5 cols
order_reviews             →  99,224 rows,  7 cols
products                  →  32,951 rows,  9 cols
sellers                   →   3,095 rows,  4 cols
geolocation               → 1,000,163 rows,  5 cols
category_translation      →      71 rows,  2 cols


### 1.2 Детальний огляд кожного файлу

Для кожної таблиці виводимо стандартний набір діагностики: розмір (`shape`), типи колонок (`dtypes`), перші 5 рядків (`head`), кількість пропусків по колонках (`isnull().sum()`), кількість повних дублікатів рядків (`duplicated().sum()`), кількість унікальних значень по кожній колонці (`nunique()`) і базову статистику (`describe(include='all')`).

Звертаємо увагу на: тип `object` у колонках з датами, несподівані мінімуми/максимуми в числових полях, різницю між кількістю рядків і кількістю унікальних значень у потенційних первинних ключах.

In [2]:
def eda_summary(df, name):
    print(f"\n{'='*60}")
    print(f"  {name}")
    print(f"{'='*60}")
    print(f"\nShape: {df.shape}")
    print(f"\nDtypes:\n{df.dtypes}")
    print(f"\nFirst 5 rows:")
    display(df.head())
    print(f"\nNull values:\n{df.isnull().sum()}")
    print(f"\nDuplicate rows: {df.duplicated().sum()}")
    print(f"\nUnique values:\n{df.nunique()}")
    print(f"\nDescribe:")
    display(df.describe(include='all'))

for name, df in dfs.items():
    eda_summary(df, name)


  customers

Shape: (99441, 5)

Dtypes:
customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

First 5 rows:


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP



Null values:
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Duplicate rows: 0

Unique values:
customer_id                 99441
customer_unique_id          96096
customer_zip_code_prefix    14994
customer_city                4119
customer_state                 27
dtype: int64

Describe:


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
count,99441,99441,99441.000000,99441,99441
unique,99441,96096,NaN,4119,27
top,06b8999e2fba1a1fbc88172c00ba8bc7,8d50f5eadf50201ccdcedfb9e2ac8455,NaN,sao paulo,SP
freq,1,17,NaN,15540,41746
mean,NaN,NaN,35137.474583,NaN,NaN
std,NaN,NaN,29797.938996,NaN,NaN
min,NaN,NaN,1003.000000,NaN,NaN
25%,NaN,NaN,11347.000000,NaN,NaN
50%,NaN,NaN,24416.000000,NaN,NaN
75%,NaN,NaN,58900.000000,NaN,NaN



  orders

Shape: (99441, 8)

Dtypes:
order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

First 5 rows:


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00



Null values:
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

Duplicate rows: 0

Unique values:
order_id                         99441
customer_id                      99441
order_status                         8
order_purchase_timestamp         98875
order_approved_at                90733
order_delivered_carrier_date     81018
order_delivered_customer_date    95664
order_estimated_delivery_date      459
dtype: int64

Describe:


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
count,99441,99441,99441,99441,99281,97658,96476,99441
unique,99441,99441,8,98875,90733,81018,95664,459
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2018-04-11 10:48:14,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-08 23:38:46,2017-12-20 00:00:00
freq,1,1,96478,3,9,47,3,522



  order_items

Shape: (112650, 7)

Dtypes:
order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

First 5 rows:


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14



Null values:
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

Duplicate rows: 0

Unique values:
order_id               98666
order_item_id             21
product_id             32951
seller_id               3095
shipping_limit_date    93318
price                   5968
freight_value           6999
dtype: int64

Describe:


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
count,112650,112650.000000,112650,112650,112650,112650.000000,112650.000000
unique,98666,NaN,32951,3095,93318,NaN,NaN
top,8272b63d03f5f79c56e9e4120aec44ef,NaN,aca2eb7d00ea1a7b8ebd4e68314663af,6560211a19b47992c3666cc44a7e94c0,2017-07-21 18:25:23,NaN,NaN
freq,21,NaN,527,2033,21,NaN,NaN
mean,NaN,1.197834,NaN,NaN,NaN,120.653739,19.990320
std,NaN,0.705124,NaN,NaN,NaN,183.633928,15.806405
min,NaN,1.000000,NaN,NaN,NaN,0.850000,0.000000
25%,NaN,1.000000,NaN,NaN,NaN,39.900000,13.080000
50%,NaN,1.000000,NaN,NaN,NaN,74.990000,16.260000
75%,NaN,1.000000,NaN,NaN,NaN,134.900000,21.150000



  order_payments

Shape: (103886, 5)

Dtypes:
order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

First 5 rows:


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45



Null values:
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Duplicate rows: 0

Unique values:
order_id                99440
payment_sequential         29
payment_type                5
payment_installments       24
payment_value           29077
dtype: int64

Describe:


,order_id,payment_sequential,payment_type,payment_installments,payment_value
count,103886,103886.000000,103886,103886.000000,103886.000000
unique,99440,NaN,5,NaN,NaN
top,fa65dad1b0e818e3ccc5cb0e39231352,NaN,credit_card,NaN,NaN
freq,29,NaN,76795,NaN,NaN
mean,NaN,1.092679,NaN,2.853349,154.100380
std,NaN,0.706584,NaN,2.687051,217.494064
min,NaN,1.000000,NaN,0.000000,0.000000
25%,NaN,1.000000,NaN,1.000000,56.790000
50%,NaN,1.000000,NaN,1.000000,100.000000
75%,NaN,1.000000,NaN,4.000000,171.837500



  order_reviews

Shape: (99224, 7)

Dtypes:
review_id                  object
order_id                   object
review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
dtype: object

First 5 rows:


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53



Null values:
review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

Duplicate rows: 0

Unique values:
review_id                  98410
order_id                   98673
review_score                   5
review_comment_title        4527
review_comment_message     36159
review_creation_date         636
review_answer_timestamp    98248
dtype: int64

Describe:


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
count,99224,99224,99224.000000,11568,40977,99224,99224
unique,98410,98673,NaN,4527,36159,636,98248
top,7b606b0d57b078384f0b58eac1d41d78,c88b1d1b157a9999ce368f218a407141,NaN,Recomendo,Muito bom,2017-12-19 00:00:00,2017-06-15 23:21:05
freq,3,3,NaN,423,230,463,4
mean,NaN,NaN,4.086421,NaN,NaN,NaN,NaN
std,NaN,NaN,1.347579,NaN,NaN,NaN,NaN
min,NaN,NaN,1.000000,NaN,NaN,NaN,NaN
25%,NaN,NaN,4.000000,NaN,NaN,NaN,NaN
50%,NaN,NaN,5.000000,NaN,NaN,NaN,NaN
75%,NaN,NaN,5.000000,NaN,NaN,NaN,NaN



  products

Shape: (32951, 9)

Dtypes:
product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

First 5 rows:


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0



Null values:
product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

Duplicate rows: 0

Unique values:
product_id                    32951
product_category_name            73
product_name_lenght              66
product_description_lenght     2960
product_photos_qty               19
product_weight_g               2204
product_length_cm                99
product_height_cm               102
product_width_cm                 95
dtype: int64

Describe:


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32951,32341,32341.000000,32341.000000,32341.000000,32949.000000,32949.000000,32949.000000,32949.000000
unique,32951,73,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,1e9e8ef04dbcff4541ed26657ea517e5,cama_mesa_banho,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,1,3029,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,NaN,NaN,48.476949,771.495285,2.188986,2276.472488,30.815078,16.937661,23.196728
std,NaN,NaN,10.245741,635.115225,1.736766,4282.038731,16.914458,13.637554,12.079047
min,NaN,NaN,5.000000,4.000000,1.000000,0.000000,7.000000,2.000000,6.000000
25%,NaN,NaN,42.000000,339.000000,1.000000,300.000000,18.000000,8.000000,15.000000
50%,NaN,NaN,51.000000,595.000000,1.000000,700.000000,25.000000,13.000000,20.000000
75%,NaN,NaN,57.000000,972.000000,3.000000,1900.000000,38.000000,21.000000,30.000000



  sellers

Shape: (3095, 4)

Dtypes:
seller_id                 object
seller_zip_code_prefix     int64
seller_city               object
seller_state              object
dtype: object

First 5 rows:


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP



Null values:
seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

Duplicate rows: 0

Unique values:
seller_id                 3095
seller_zip_code_prefix    2246
seller_city                611
seller_state                23
dtype: int64

Describe:


,seller_id,seller_zip_code_prefix,seller_city,seller_state
count,3095,3095.000000,3095,3095
unique,3095,NaN,611,23
top,3442f8959a84dea7ee197c632cb2df15,NaN,sao paulo,SP
freq,1,NaN,694,1849
mean,NaN,32291.059451,NaN,NaN
std,NaN,32713.453830,NaN,NaN
min,NaN,1001.000000,NaN,NaN
25%,NaN,7093.500000,NaN,NaN
50%,NaN,14940.000000,NaN,NaN
75%,NaN,64552.500000,NaN,NaN



  geolocation

Shape: (1000163, 5)

Dtypes:
geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                object
geolocation_state               object
dtype: object

First 5 rows:


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP



Null values:
geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

Duplicate rows: 261831

Unique values:
geolocation_zip_code_prefix     19015
geolocation_lat                717363
geolocation_lng                717615
geolocation_city                 8011
geolocation_state                  27
dtype: int64

Describe:


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
count,1.000163e+06,1.000163e+06,1.000163e+06,1000163,1000163
unique,NaN,NaN,NaN,8011,27
top,NaN,NaN,NaN,sao paulo,SP
freq,NaN,NaN,NaN,135800,404268
mean,3.657417e+04,-2.117615e+01,-4.639054e+01,NaN,NaN
std,3.054934e+04,5.715866e+00,4.269748e+00,NaN,NaN
min,1.001000e+03,-3.660537e+01,-1.014668e+02,NaN,NaN
25%,1.107500e+04,-2.360355e+01,-4.857317e+01,NaN,NaN
50%,2.653000e+04,-2.291938e+01,-4.663788e+01,NaN,NaN
75%,6.350400e+04,-1.997962e+01,-4.376771e+01,NaN,NaN



  category_translation

Shape: (71, 2)

Dtypes:
product_category_name            object
product_category_name_english    object
dtype: object

First 5 rows:


,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor



Null values:
product_category_name            0
product_category_name_english    0
dtype: int64

Duplicate rows: 0

Unique values:
product_category_name            71
product_category_name_english    71
dtype: int64

Describe:


,product_category_name,product_category_name_english
count,71,71
unique,71,71
top,beleza_saude,health_beauty
freq,1,1


### 1.3 Перевірка первинних ключів

Перевіряємо унікальність первинних ключів у кожній таблиці. Для `order_items` — складений PK `(order_id, order_item_id)`, тому перевіряємо унікальність комбінації. Для `order_payments` — множинні платежі на одне замовлення є нормою, тому виводимо статистику без перевірки на унікальність `order_id`.

In [3]:
pk_checks = {
    'customers':            'customer_id',
    'orders':               'order_id',
    'products':             'product_id',
    'sellers':              'seller_id',
    'order_reviews':        'review_id',
    'category_translation': 'product_category_name',
}

for table, pk in pk_checks.items():
    is_unique = dfs[table][pk].is_unique
    status = '' if is_unique else ''
    print(f"{status} {table}.{pk} — unique: {is_unique}")

 customers.customer_id — unique: True
 orders.order_id — unique: True
 products.product_id — unique: True
 sellers.seller_id — unique: True
 order_reviews.review_id — unique: False
 category_translation.product_category_name — unique: True


In [4]:
# order_items — складений ключ (order_id + order_item_id)
dupes = dfs['order_items'].duplicated(subset=['order_id', 'order_item_id']).sum()
print(f"{'' if dupes == 0 else ''} order_items (order_id + order_item_id) — duplicates: {dupes}")

# order_payments — може мати кілька платежів на замовлення (це нормально)
print(f" order_payments — {dfs['order_payments']['order_id'].nunique()} unique orders, "
      f"{len(dfs['order_payments'])} rows (multiple payments per order expected)")

 order_items (order_id + order_item_id) — duplicates: 0
 order_payments — 99440 unique orders, 103886 rows (multiple payments per order expected)


### 1.4 Попередні спостереження

На основі кроків 1.2 та 1.3 зафіксовано наступні спостереження, які потребують поглибленої перевірки:

**Типи даних:**
- У таблицях `orders`, `order_items`, `order_reviews` усі колонки з датами мають тип `object` замість `datetime`. Це заважатиме розрахункам (час доставки, динаміка по місяцях). Потребує конвертації на етапі ETL.

**Пропуски (NULL):**
- `orders`: колонки `order_approved_at` (160), `delivered_carrier_date` (1783), `delivered_customer_date` (2965) мають пропуски. Потрібно перевірити — чи це лише незавершені замовлення (статус ≠ delivered), чи є аномалії серед доставлених.
- `order_reviews`: ~88% пропусків у `comment_title`, ~59% у `comment_message`. На перший погляд це нормально (не всі пишуть текст), але потрібно підтвердити.
- `products`: 610 пропусків одночасно в `category_name`, `name_lenght`, `description_lenght`, `photos_qty` — схоже, це ті самі 610 рядків з неповним заповненням. Окремо 2 пропуски у фізичних розмірах (weight, length, height, width).

**Первинні ключі:**
- `order_reviews.review_id` — NOT unique (98410 унікальних з 99224 рядків). Це критична проблема, бо за планом `review_id` є PK. Потрібно зрозуміти: це повні дублікати рядків чи різні відгуки з випадково однаковим ID?

**Підозрілі значення в describe():**
- `order_items`: `price` min=0.85, max=6735 при mean=120 — потенційні outliers у верхній частині. `freight_value` min=0.00 — є товари з безкоштовною доставкою, потрібно перевірити масштаб.
- `order_payments`: `payment_installments` min=0 — розстрочка на 0 платежів нелогічна. `payment_value` min=0.00 — нульові платежі потребують пояснення. Max 29 послідовних платежів на одне замовлення — можливий edge case.
- `products`: `product_weight_g` min=0 — товар з нульовою вагою нелогічний для фізичної доставки.

**Дублікати:**
- `geolocation`: 261831 повний дублікат (~26%). Також у describe() видно координати за межами Бразилії (lat max=+45, lng max=+121) — потрібно перевірити масштаб проблеми.

**Розбіжність довідників:**
- У `products` є 73 унікальні категорії, а в `category_translation` лише 71 переклад. 2 категорії залишаться без англійської назви після merge.

**Зв'язки між таблицями:**
- `customers` має 99441 `customer_id`, але лише 96096 `customer_unique_id` — один реальний клієнт може мати кілька технічних ID. Це впливає на розрахунок repeat rate (Питання 6 в аналізі).
- FK-цілісність поки не перевірена — потрібно переконатися, що всі `product_id`, `seller_id`, `order_id` в дочірніх таблицях мають відповідники в батьківських.

Далі проводимо поглиблену перевірку кожного з цих спостережень.

### 1.5 Поглиблена перевірка знайдених проблем

Перевіряємо кожне підозріле спостереження зі секції 1.4: NULL у датах доставлених замовлень, нульові значення в числових полях, дублікати у відгуках, нестандартні координати в geolocation та FK-цілісність між таблицями. Для кожної проблеми визначаємо масштаб і підтверджуємо рішення.

In [5]:
# orders — NULL у датах доставки для статусу 'delivered'
print(f"\nDelivered без approved_at: "
      f"{dfs['orders'][(dfs['orders']['order_status']=='delivered') & (dfs['orders']['order_approved_at'].isnull())].shape[0]}")
print(f"Delivered без delivered_customer_date: "
      f"{dfs['orders'][(dfs['orders']['order_status']=='delivered') & (dfs['orders']['order_delivered_customer_date'].isnull())].shape[0]}")

# orders — розподіл статусів (скільки не delivered)
print(dfs['orders']['order_status'].value_counts())


Delivered без approved_at: 14
Delivered без delivered_customer_date: 8
order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64


In [6]:
# order_items — нульовий freight і нульова/мінімальна ціна
print(f"price <= 0: {(dfs['order_items']['price'] <= 0).sum()}")
print(f"freight_value == 0: {(dfs['order_items']['freight_value'] == 0).sum()}")

price <= 0: 0
freight_value == 0: 383


In [7]:
# order_payments — нульові installments і нульовий value
print(f"installments == 0: {(dfs['order_payments']['payment_installments'] == 0).sum()}")
print(f"value == 0: {(dfs['order_payments']['payment_value'] == 0).sum()}")

# Які типи платежів мають 0 installments?
print(f"\nТипи платежів з 0 installments:")
print(dfs['order_payments'][dfs['order_payments']['payment_installments'] == 0]['payment_type'].value_counts())

installments == 0: 2
value == 0: 9

Типи платежів з 0 installments:
payment_type
credit_card    2
Name: count, dtype: int64


In [8]:
# order_reviews — дублікати review_id
review_dupes = dfs['order_reviews'][dfs['order_reviews'].duplicated(subset='review_id', keep=False)]
print(f"Рядків з дубльованим review_id: {len(review_dupes)}")
display(review_dupes.sort_values('review_id').head(10))

Рядків з дубльованим review_id: 1603


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
46678,00130cbe1f9d422698c812ed8ded1919,dfcdfc43867d1c1381bfaf62d6b9c195,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07 00:00:00,2018-03-20 18:08:23
29841,00130cbe1f9d422698c812ed8ded1919,04a28263e085d399c97ae49e0b477efa,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07 00:00:00,2018-03-20 18:08:23
90677,0115633a9c298b6a98bcbe4eee75345f,78a4201f58af3463bdab842eea4bc801,5,NaN,NaN,2017-09-21 00:00:00,2017-09-26 03:27:47
63193,0115633a9c298b6a98bcbe4eee75345f,0c9850b2c179c1ef60d2855e2751d1fa,5,NaN,NaN,2017-09-21 00:00:00,2017-09-26 03:27:47
92876,0174caf0ee5964646040cd94e15ac95e,f93a732712407c02dce5dd5088d0f47b,1,NaN,Produto entregue dentro de embalagem do fornec...,2018-03-07 00:00:00,2018-03-08 03:00:53
57280,0174caf0ee5964646040cd94e15ac95e,74db91e33b4e1fd865356c89a61abf1f,1,NaN,Produto entregue dentro de embalagem do fornec...,2018-03-07 00:00:00,2018-03-08 03:00:53
54832,017808d29fd1f942d97e50184dfb4c13,8daaa9e99d60fbba579cc1c3e3bfae01,5,NaN,NaN,2018-03-02 00:00:00,2018-03-05 01:43:30
99167,017808d29fd1f942d97e50184dfb4c13,b1461c8882153b5fe68307c46a506e39,5,NaN,NaN,2018-03-02 00:00:00,2018-03-05 01:43:30
20621,0254bd905dc677a6078990aad3331a36,5bf226cf882c5bf4247f89a97c86f273,1,NaN,O pedido consta de 2 produtos e até agora rece...,2017-09-09 00:00:00,2017-09-13 09:52:44
96080,0254bd905dc677a6078990aad3331a36,331b367bdd766f3d1cf518777317b5d9,1,NaN,O pedido consta de 2 produtos e até agora rece...,2017-09-09 00:00:00,2017-09-13 09:52:44


In [9]:
# products — нульова вага
print(f"weight == 0: {(dfs['products']['product_weight_g'] == 0).sum()}")

# products — категорії без перекладу
cats_in_products = set(dfs['products']['product_category_name'].dropna().unique())
cats_in_translation = set(dfs['category_translation']['product_category_name'].unique())
print(f"\nКатегорії без перекладу: {cats_in_products - cats_in_translation}")

weight == 0: 4

Категорії без перекладу: {'portateis_cozinha_e_preparadores_de_alimentos', 'pc_gamer'}


In [10]:
# geolocation — координати за межами Бразилії
geo = dfs['geolocation']
outside = geo[(geo['geolocation_lat'] < -34) | (geo['geolocation_lat'] > 6) |
              (geo['geolocation_lng'] < -74) | (geo['geolocation_lng'] > -34)]
print(f"Точок за межами Бразилії: {len(outside)} ({len(outside)/len(geo)*100:.3f}%)")

Точок за межами Бразилії: 42 (0.004%)


In [11]:
# FK-валідація — чи всі order_items.product_id існують у products
orphan_products = set(dfs['order_items']['product_id']) - set(dfs['products']['product_id'])
print(f"order_items з неіснуючим product_id: {len(orphan_products)}")

orphan_sellers = set(dfs['order_items']['seller_id']) - set(dfs['sellers']['seller_id'])
print(f"order_items з неіснуючим seller_id: {len(orphan_sellers)}")

orphan_orders = set(dfs['order_items']['order_id']) - set(dfs['orders']['order_id'])
print(f"order_items з неіснуючим order_id: {len(orphan_orders)}")

orphan_reviews = set(dfs['order_reviews']['order_id']) - set(dfs['orders']['order_id'])
print(f"order_reviews з неіснуючим order_id: {len(orphan_reviews)}")

order_items з неіснуючим product_id: 0
order_items з неіснуючим seller_id: 0
order_items з неіснуючим order_id: 0
order_reviews з неіснуючим order_id: 0


### 1.6 Зведення знайдених проблем

| # | Файл | Проблема | Критичність | Рішення (Крок 3) | Деталі |
|---|------|----------|-------------|-------------------|--------|
| 1 | orders | Усі дати як `object` | Висока | `pd.to_datetime()` | 5 колонок з датами |
| 2 | orders | NULL у датах доставлених замовлень | Низька | Залишити — масштаб не впливає на аналіз | 14 delivered без approved_at, 8 без delivered_customer_date з 96478 |
| 3 | orders | ~3% замовлень не delivered | Низька | Зафіксувати; для аналізу доставки фільтрувати по status=delivered | shipped: 1107, canceled: 625, unavailable: 609, invoiced: 314, processing: 301, created: 5, approved: 2 |
| 4 | order_items | Дата `shipping_limit_date` як `object` | Висока | `pd.to_datetime()` | — |
| 5 | order_items | `freight_value` = 0 у 383 записах | Низька | Не потребує дій — безкоштовна доставка або самовивіз | price <= 0 відсутні (добре) |
| 6 | order_payments | `payment_installments` = 0 | Низька | Залишити або видалити 2 записи | Обидва — credit_card, мізерний масштаб |
| 7 | order_payments | `payment_value` = 0 | Низька | Залишити або видалити 9 записів | Технічні записи |
| 8 | order_payments | До 29 платежів на одне замовлення | Низька | Залишити — edge case | order_id `fa65dad...` має freq=29 |
| 9 | order_reviews | `review_id` не унікальний | Середня | Складений PK: `(review_id, order_id)` | 1603 рядки; один відгук прив'язаний до різних order_id того самого клієнта — не помилка, а особливість даних |
| 10 | order_reviews | NULL у коментарях | Низька | Залишити NULL | comment_title: 87656 (~88%), comment_message: 58247 (~59%) — не всі пишуть текст |
| 11 | order_reviews | Дати як `object` | Висока | `pd.to_datetime()` | review_creation_date, review_answer_timestamp |
| 12 | products | NULL у category + метаданих | Середня | Категорію → 'unknown', решта — залишити NULL | 610 рядків без category_name, name_lenght, description_lenght, photos_qty |
| 13 | products | NULL у фізичних розмірах | Низька | Заповнити медіаною або видалити | weight_g, length_cm, height_cm, width_cm: по 2 NULL |
| 14 | products | `product_weight_g` = 0 | Низька | Заповнити медіаною по категорії | 4 товари з нульовою вагою |
| 15 | products | 2 категорії без перекладу | Середня | Додати вручну | `pc_gamer` → 'pc_gamer', `portateis_cozinha_e_preparadores_de_alimentos` → 'portable_kitchen_food_processors' |
| 16 | geolocation | 261831 повних дублікатів | Середня | `drop_duplicates()` | ~26% від 1M рядків |
| 17 | geolocation | Координати за межами Бразилії | Низька | Фільтр: lat ∈ [-34, 6], lng ∈ [-74, -34] | Лише 42 точки (0.004%) — мінімальний масштаб |
| 18 | customers | `customer_unique_id` не унікальний | Низька | Для repeat rate використовувати `customer_unique_id` | 96096 унікальних vs 99441 customer_id |
| 19 | FK-зв'язки | Цілісність зв'язків | — | Не потребує дій | 0 orphans по всіх FK (order_items → products/sellers/orders, reviews → orders) |

---

### Висновок

В EDA було виявлено **19 спостережень** у 7 таблицях. Критичних проблем, які унеможливлюють аналіз, немає — датасет придатний до роботи після очищення.

**Ключові рішення для ETL (`02_etl.ipynb`):**
- Конвертувати дати з `object` → `datetime` у таблицях `orders`, `order_items`, `order_reviews` (5+2+2 колонки).
- Заповнити NULL у фізичних розмірах товарів медіаною по категорії; `product_category_name` NULL → `'unknown'`.
- Додати переклади для 2 категорій без відповідника в `category_translation`: `pc_gamer` і `portateis_cozinha_e_preparadores_de_alimentos`.
- Дедублікувати `order_reviews` по складеному ключу `(review_id, order_id)` — зберегти перший запис.
- Агрегувати `geolocation` медіаною координат по zip-коду: ~1M рядків → ~15k унікальних zip.
- Малозначні аномалії (freight=0, installments=0, value=0, 42 точки поза Бразилією) — залишити або видалити на рівні ETL відповідно до бізнес-логіки.

FK-цілісність між усіма таблицями підтверджена: 0 orphans. Дані готові до завантаження в PostgreSQL.